diff --git a//wasm/Writer.cpp b//wasm/Writer.cpp
--- a//wasm/Writer.cpp	(revision 320191)
+++ b//wasm/Writer.cpp	(working copy)
@@ -252,11 +252,18 @@
   bool ExportEntry = !Config->Relocatable && EntrySym && EntrySym->isDefined();
   bool ExportHidden = Config->EmitRelocs;
 
+  llvm::SmallDenseSet<StringRef, 4> ExportNames;
+
   uint32_t NumExports = ExportMemory ? 1 : 0;
+  if (ExportMemory) {
+    ExportNames.insert("memory");
+  }
 
   std::vector<const Symbol *> SymbolExports;
-  if (ExportEntry)
+  if (ExportEntry) {
     SymbolExports.emplace_back(EntrySym);
+    ExportNames.insert(EntrySym->getName());
+  }
 
   for (const Symbol *Sym : Symtab->getSymbols()) {
     if (Sym->isUndefined() || Sym->isGlobal())
@@ -265,7 +272,10 @@
       continue;
     if (ExportEntry && Sym == EntrySym)
       continue;
+    if (ExportNames.find(Sym->getName()) != ExportNames.end())
+      continue;
     SymbolExports.emplace_back(Sym);
+    ExportNames.insert(Sym->getName());
   }
 
   for (const Symbol *Sym : DefinedGlobals) {
@@ -273,7 +283,10 @@
     // connot be exported.
     if (Sym == Config->StackPointerSymbol)
       continue;
+    if (ExportNames.find(Sym->getName()) != ExportNames.end())
+      continue;
     SymbolExports.emplace_back(Sym);
+    ExportNames.insert(Sym->getName());
   }
 
   NumExports += SymbolExports.size();
