diff --git a//lib/MC/WasmObjectWriter.cpp b//lib/MC/WasmObjectWriter.cpp
--- a//lib/MC/WasmObjectWriter.cpp	(revision 319728)
+++ b//lib/MC/WasmObjectWriter.cpp	(working copy)
@@ -1294,9 +1294,16 @@
     WasmExport Export;
     Export.FieldName = WS.getName();
     Export.Index = Index;
-    if (WS.isFunction())
+    if (WS.isFunction()) {
+      // If needed, prepare the function to be called indirectly.
+      if (IsAddressTaken.count(&WS) != 0) {
+        IndirectSymbolIndices[&WS] = TableElems.size();
+        DEBUG(dbgs() << "  -> adding to table: (hack) " << TableElems.size() << "\n");
+        TableElems.push_back(Index);
+      }
+
       Export.Kind = wasm::WASM_EXTERNAL_FUNCTION;
-    else
+    } else
       Export.Kind = wasm::WASM_EXTERNAL_GLOBAL;
     DEBUG(dbgs() << "  -> export " << Exports.size() << "\n");
     Exports.push_back(Export);
