Index: include/llvm/BinaryFormat/Wasm.h
===================================================================
diff --git a/llvm/trunk/include/llvm/BinaryFormat/Wasm.h b/llvm/trunk/include/llvm/BinaryFormat/Wasm.h
--- a/llvm/trunk/include/llvm/BinaryFormat/Wasm.h	(revision 331690)
+++ b/llvm/trunk/include/llvm/BinaryFormat/Wasm.h	(working copy)
@@ -102,6 +102,7 @@
   ArrayRef<uint8_t> Body;
   uint32_t CodeSectionOffset;
   uint32_t Size;
+  uint32_t FunctionContentOffset; // aka Size field size
   StringRef SymbolName; // from the "linking" section
   StringRef DebugName; // from the "name" section
   uint32_t Comdat; // from the "comdat info" section
Index: lib/Object/WasmObjectFile.cpp
===================================================================
diff --git a/llvm/trunk/lib/Object/WasmObjectFile.cpp b/llvm/trunk/lib/Object/WasmObjectFile.cpp
--- a/llvm/trunk/lib/Object/WasmObjectFile.cpp	(revision 331690)
+++ b/llvm/trunk/lib/Object/WasmObjectFile.cpp	(working copy)
@@ -875,6 +875,7 @@
     uint32_t Size = readVaruint32(Ptr);
     const uint8_t *FunctionEnd = Ptr + Size;
 
+    Function.FunctionContentOffset = Ptr - FunctionStart;
     Function.Index = NumImportedFunctions + Functions.size();
     Function.CodeSectionOffset = FunctionStart - CodeSectionStart;
     Function.Size = FunctionEnd - FunctionStart;
